src/Lecture1.hs

```diff
@@ -98,7 +105,15 @@ start position can be considered as zero (e.g. substring from the
 first character) and negative end position should result in an empty
 string.
 -}
-subString start end str = error "TODO"
+subString :: Int -> Int -> String -> String
+subString start end str = take numOfCharsToKeep $ drop start str
+    where numOfCharsToKeep = end' - start' + 1
+          start' = if (start < 0)
+                      then 0
+                      else start
+          end' = if (end < 0 || end < start)
+                    then start - 1
+                    else end
```

Comment:
It's a clever solution and I can see how by defining `end` lower than `start` you can achieve `numOfCharsToKeep` equal to zero.

But this code is a bit hard to follow. For clarity, I would use guards in this function. Something like this:

```haskell
subString start end str
    | end < 0     = ""
    | end < start = ""
    | otherwise   = ...
  where
    ...
```